File: matlab/criterion.m
function mae = criterion(pnlAcctSoFar, cumPnl, af, option)

switch option
    
    case '0'
        
        mae = 0;
        nAccts = length(pnlAcctSoFar);
        for i = 1:nAccts
            for j = 1:nAccts
                if i~=j
                    mae = mae+(pnlAcctSoFar(i)/af(i)-pnlAcctSoFar(j)/af(j))^2;
                end
            end
        end       
    
    case '1'
        
        mae = sum((pnlAcctSoFar./af - cumPnl).^2);
        
    case '2'
        
        mae = sum(abs((pnlAcctSoFar - af.*cumPnl)./af.*cumPnl));
        
    case '3'
        
        mae = sum((pnlAcctSoFar - af.*cumPnl).^2);
        
    case '4'
        
        mae = sum(abs(pnlAcctSoFar - af.*cumPnl));
        
end

File: criterion.cpp
/*
* @Author: shayaan
* @Date:   2020-06-12 11:48:37
* @Last Modified by:   shayaan
* @Last Modified time: 2020-06-15 15:52:53
*/


#include "allInclude.h"

static double criterion(vector<double> pnlAcctSoFar, double cumPnl, vector<double> af, int option){
    double mae = 0;
    double nAccts = pnlAcctSoFar.size();
    switch(option){
       case 0:
           for(int i = 0;i < nAccts; i++){
               for(int j = 0;j < nAccts; j++){
                   if (i!=j){
                       mae += (pnlAcctSoFar[i]/af[i]-pnlAcctSoFar[j]/af[j])*(pnlAcctSoFar[i]/af[i]-pnlAcctSoFar[j]/af[j]);
                   }
               }
           }
           break;
        case 1:
            for(int i = 0;i < nAccts; i++){
                double incr = pnlAcctSoFar[i]/af[i] - cumPnl;
                mae += (incr*incr);
            }
            break;

       case 2:
           for(int i = 0;i < nAccts; i++){
               double incr = fabs((pnlAcctSoFar[i]-af[i]*cumPnl)/af[i]*cumPnl);
               mae += incr;
           };
           break;
       case 3:
           for(int i = 0;i < nAccts; i++){
               double incr = (pnlAcctSoFar[i]-af[i]*cumPnl);
               mae += incr*incr;
           };
           break;

       case 4:
           for(int i = 0;i < nAccts; i++){
               double incr = fabs(pnlAcctSoFar[i]-af[i]*cumPnl);
               mae += incr;
           };
           break;

    }
    return mae;
}